xknxproject/models/knxproject.py

```diff
@@ -92,6 +92,18 @@ class Space(TypedDict):
     project_uid: int | None
     devices: list[str]
     spaces: dict[str, Space]
+    functions: list[str, Functions]
+
+
+class Functions(TypedDict):
+    """Functions typed dict."""
+
+    identifier: str
+    name: str
+    function_type: str
+    project_uid: int | None
+    group_addresses: list[str]
```

Comment:
Is it possible to distinguish between GA-roles? 
Eg. for a light we'd probably want to know what is a status address vs what is a command address and also what is the switch and what is the brightness...

Reply:
I don't think that's so easy, you can only parse the name and compare it. Here I can also create examples with a dimmable lamp and a blind.

I think that's the job of the gui. I think the alias generator in io-broker does something similar, but only looks at whether it is a status or control object.


Another problem is that you have to split the group address at _ and then look for the identifier GA-1 in the group_addresses to get the group address.

That's the interesting part from the generated XML.
```
<Space Type="Room" Usage="SU-4" Id="P-05C0-0_BP-2" Name="Livingroom" Puid="10">
    <Function Id="P-05C0-0_F-1" Name="LivingroomLight" Type="FT-1" Puid="11">
       <GroupAddress Id="P-05C0-0_GA-1" Address="1" Name="Schalten" Description="Livingroom LivingroomLight" DatapointType="DPST-1-1" Puid="14" />
       <GroupAddress Id="P-05C0-0_GA-2" Address="2" Name="Status" Description="Livingroom LivingroomLight" DatapointType="DPST-1-1" Puid="16" /
   </Function>
</Space>
```

Reply:
Have a look at the `knx_master.xml` file. The `FunctionType`s and `FunctionPoint`s are defined there. You should also find the translated names of the function types there.

The group addresses are prefixed with the project_id in some places in `0.xml` and in some not(eg. ComObjectInstanceRef) 🤷 I don't know why... 

Unfortunately I don't know io-broker so can't really make anything of the comparison.

Reply:
I had copied the wrong part from the XML. Now I could extend the group_addresses with Role, Name, Id and Puid.

If it is possible to access the knx_master_loader at the time of the space parser, then the function_type could still be resolved to a readable type. Or would I have to implement that in xml/parser.py?

Reply:
We should avoid reading the file twice. You can just parse out all `FunctionType` tags from `knx_master.xml` into a dict when it is loaded and use the ones you find in the project file later. Just like we already do with `space_usage_mapping` https://github.com/XKNX/xknxproject/blob/115ab8ce6f6ad96e2bce64cbfd023e0b61d9a58c/xknxproject/xml/parser.py#L188
In Schema-version 20 there are only 20 FunctionType tags, but I guess this will become more with future versions.

Reply:
That worked. I have now added the usage_text and the address.